crates/pathfinder/src/cairo/ext_py/sub_process.rs

```diff
@@ -0,0 +1,318 @@
+    let sleep = tokio::time::sleep(std::time::Duration::from_millis(1000));
+    tokio::pin!(sleep);
+
+    let exit_status = tokio::select! {
+        _ = &mut sleep => {
+            match child.kill().await {
+                Ok(()) => {}
+                Err(e) => println!("Killing python subprocess failed, possibly a race? {e:?}"),
+            }
+
+            // kill already await the child, so there's not much to await here, we should just get the
+            // fused response.
+            match child.wait().await {
+                Ok(status) => Some(status),
+                Err(e) => {
+                    eprintln!("wait on child pid failed: {e:?}");
+                    None
+                }
+            }
+        }
+        exit_status = child.wait() => {
+            exit_status.ok()
+        }
+    };
+
+    Some((pid, exit_status, exit_reason))
+}
+
+async fn spawn(
+    database_path: PathBuf,
+) -> anyhow::Result<(Child, u32, ChildStdin, BufReader<ChildStdout>, String)> {
+    // there is not intentionally any std::fs::exists calls to avoid bringing any TOCTOU issues.
+    let virtual_env = std::env::var_os("VIRTUAL_ENV")
+        .context("VIRTUAL_ENV is not defined; has the user activated virtual environment")?;
+
+    // FIXME: use choom, add something over /proc/self/oom_score_adj ?
+    let mut python_exe = PathBuf::from(&virtual_env);
+    python_exe.push("bin");
+    python_exe.push("python");
+
+    // we assume that VIRTUAL_ENV is at the base of the `py/` directory
+    let mut call_py = PathBuf::from(&virtual_env);
+    call_py.push("..");
+    call_py.push("src");
+    call_py.push("call.py");
+
+    let mut command = tokio::process::Command::new(python_exe);
+
+    command
+        .arg(call_py)
+        .arg(database_path)
+        .env("VIRTUAL_ENV", virtual_env)
+        .stdin(std::process::Stdio::piped())
+        .stdout(std::process::Stdio::piped())
+        .kill_on_drop(true);
+
+    let _inherit_stderr = false;
+
+    #[cfg(debug_assertions)]
+    let _inherit_stderr = true;
+
+    if _inherit_stderr {
+        // in case the file doesn't exist it would be reported here ... probably cannot leave it
+        // like this. exit code 2 is for the python file not being found.
+        //
+        // debug info is a great upside, but for example CTRL-C gets to the child processses as
+        // well, and python will print a stacktrace when an SIGINT happens, which might be
+        // confusing for users, who had just ran a rust program.
+        command.stderr(std::process::Stdio::inherit());
+    } else {
+        command.stderr(std::process::Stdio::null());
+    }
+
+    let mut child = command
+        .spawn()
+        .context("Failed to spawn the new python process; this should only happen when the session is at it's process limit on unix.")?;
+
+    let pid = match child.id() {
+        Some(pid) => pid,
+        None => todo!("child spawned but exited before reading pid"),
```

Comment:
```suggestion
        None => {
            // I've understood that this should not happen as long as we haven't `wait`'d the child
            // because that is the only way we could learn that the pid has exited and the `id()` should
            // start return `None`.
            todo!("child spawned but exited before reading pid")
        },
```

Or something, maybe just an `.expect` with a longer message instead of a comment.

Reply:
Moving to expect, adding comment above.